src/rgw/radosgw-admin/orphan.cc

```diff
@@ -1060,24 +1057,35 @@ int RGWRadosList::pop_and_handle_stat_op(
   RGWObjectCtx& obj_ctx,
   std::deque<RGWRados::Object::Stat>& ops)
 {
-  std::string bucket_name;
-  rgw_obj_key obj_key;
   std::set<std::string> obj_oids;
   RGWRados::Object::Stat& front_op = ops.front();
 
   int ret = front_op.wait(dpp);
-  if (ret < 0) {
-    if (ret != -ENOENT) {
-      ldpp_dout(dpp, -1) << "ERROR: stat_async() returned error: " <<
-	cpp_strerror(-ret) << dendl;
-    }
-    goto done;
-  }
+  // note: even if we get an error, front_op.result.obj will still be
+  // populated with correct data
+  const std::string bucket_name = front_op.result.obj.bucket.name;
+  const rgw_obj_key obj_key = front_op.result.obj.key;
 
-  ret = handle_stat_result(dpp, front_op.result, bucket_name, obj_key, obj_oids);
-  if (ret < 0) {
-    ldpp_dout(dpp, -1) << "ERROR: handle_stat_result() returned error: " <<
+  if (ret == -ENOENT) {
+    const auto& result = front_op.result;
+    const rgw_bucket& bucket = result.obj.bucket;
+    const std::string oid = bucket.marker + "_" + result.obj.get_oid();
+    obj_oids.insert(oid);
```

Comment:
looks correct for regular objects. multipart head objects seem to require special handling in `handle_stat_result()`. should we probably add logic to handle multipart here in this ENOENT block as well?

Reply:
I tested it with completed multipart objects. I tried removing a rados object in the _shadow_ namespace, _multipart_ namespace, and the head object, and in each case radoslist and rgw-gap-list worked as expected.

Did you mean incomplete multipart uploads?

Reply:
Incomplete multipart uploads work also.

The area of the if/else if/else is for the case when the head object cannot be read. The else area handles when the head object is read. It then enters `handle_stat_result` where it processes the manifest to determine all the tail objects.

Incomplete multipart uploads are handled at the end of `RGWRadosList::run(DoutPrefixProvider, string, bool)` and `RGWRadosList::do_incomplete_multipart`.

Reply:
got it. thanks for confirming!